### PR TITLE
Add helper functions to aid in migration of AsdfInFits from ASDF

### PR DESCRIPTION
**Pull request by braingram**
This PR adds an asdf_in_fits submodule with 2 functions:
- open
- write

These are intended as near replacements for:
- opening an AsdfInFits file
- writing an AsdfInFits file

The signatures of the replacement functions were determined by surveying AsdfInFits usage outside of stdatamodels/jwst. These uses include:
- [hcipy](https://github.com/ehpor/hcipy)
  - read: call to [`asdf.fits_embed.AsdfInFits.open(filename)`](https://github.com/ehpor/hcipy/blob/058c473b31ae094f44ad1a3e445f6773c18f0d8a/hcipy/util/io.py#L140)
  - write: call to [`ff = asdf.fits_embed.AsdfInFits(hdulist, tree); ff.write_to(filename, …)`](https://github.com/ehpor/hcipy/blob/058c473b31ae094f44ad1a3e445f6773c18f0d8a/hcipy/util/io.py#L187-L188)
- [specutils](https://github.com/astropy/specutils)
  - read: call to [`with asdf.open(filename) as af`](https://github.com/astropy/specutils/blob/fc04cd6c948698c70ec9bd15b3354b58e4dbbcc2/specutils/io/default_loaders/jwst_reader.py#L438-L452)
  - read: call to [`with asdf.open(filename) as af`](https://github.com/astropy/specutils/blob/fc04cd6c948698c70ec9bd15b3354b58e4dbbcc2/specutils/io/default_loaders/jwst_reader.py#L558-L559)
  - write: call to [`ff = asdf.fits_embed.AsdfInFits(hdulist, tree); ff.write_to(filename, …)`](https://github.com/astropy/specutils/blob/fc04cd6c948698c70ec9bd15b3354b58e4dbbcc2/specutils/io/default_loaders/tests/test_jwst_reader.py#L409)
- [jdaviz](https://github.com/spacetelescope/jdaviz/)
  - read: call to [`with AsdfInFits.open as af`](https://github.com/spacetelescope/jdaviz/blob/20b224e616f61f0fd4245807560d7025aefc6867/jdaviz/configs/imviz/plugins/parsers.py#L232-L234)
  - write: call to [`ff = asdf.fits_embed.AsdfInFits(hdulist, tree); ff.write_to(filename, …)`](https://github.com/spacetelescope/jdaviz/blob/f257d716a101b59807b4dc05a1e4dd19c020d30b/jdaviz/configs/imviz/tests/test_parser.py#L230-L231)
 
To summarize, the above uses include opening as a with context or as a function call and writing by creating an `AsdfInFits` object then calling `write_to`.

For reads using a with statement, calls like the following:
```python
with asdf.open(filename) as af:  # or AsdfInFits.open
```
can be replaced with
```python
with stdatmodels.asdf_in_fits.open(filename) as af:
```

The same function can be used outside of a with statement to allow replacing:
```python
af = asdf.open(filename)
```
with
```python
af = stdatamodels.asdf_in_fits.open(filename)
```

Finally writes like the following:
```python
hdulist = fits.HDUList()
ff = asdf.fits_embed.AsdfInFits(hdulist, tree)
ff.write_to(filename, …)
```
can be replaced with
```python
stdatamodels.asdf_in_fits.write(filename, tree)
```
This and several related forms (including passing in an empty or partially full hdu) is included as tests.

The rendering of the documentation appears odd as the numpy style docstrings don't appear to correctly render locally (there is no space between parameter names and types).

**Checklist**
- [ ] added entry in `CHANGES.rst` (either in `Bug Fixes` or `Changes to API`)
- [ ] updated relevant tests
- [ ] updated relevant documentation
- [ ] updated relevant milestone(s)
- [ ] added relevant label(s)

### CODE REVIEW

File: src/stdatamodels/fits_support.py [hunk 1]
```diff
@@ -897,11 +907,88 @@ def get_short_doc(schema):
...
+#
+# Read:
+#     ff = asdf.open(filename)  # where filename is a fits file
+#       -- or explicity --
+#     ff = asdf.fits_embed.AsdfInFits.open(filename)
+#
+# Read in context:
+#     with asdf.open(filename) as af:
+#       -- or explicitly --
+#     with asdf.fits_embed.AsdfInFits.open(filename) as af:
```
Comment (WilliamJamieson) [inline src/stdatamodels/fits_support.py:939]: If this part of the module needs this header, then it should be its own module with this as the module's docstring.

**Review by WilliamJamieson — CHANGES_REQUESTED** [review]
This should have an actual docs page. Indeed, `fits_support` in general should have a detailed documentation page anyways, so that we can link to it from the ASDF documentation.

**Comment by braingram** [comment]
JWST regression tests passed:
https://plwishmaster.stsci.edu:8081/job/RT/job/JWST/2204/

File: src/stdatamodels/fits_support.py (hunk 1 above)
Comment (braingram) [reply src/stdatamodels/fits_support.py]: These were leftover comments from my own notes. I've removed them in favor of including something similar in the actual docs.

**Comment by braingram** [comment]
> This should have an actual docs page. Indeed, `fits_support` in general should have a detailed documentation page anyways, so that we can link to it from the ASDF documentation.

Are the docs for this package published anywhere?

Adding documentation to fits_support seems outside the scope of this PR.

What is your opinion on adding the migration documentation to ASDF as it has more thorough documentation and these functions are being added to aid in migration for ASDF users?

**Comment by WilliamJamieson** [comment]
> Are the docs for this package published anywhere?
> 

It should @zacharyburnett or @nden should know where it is (or we should create a RTD). There is a `docs` in this repo which is functional, though it does not contain much.

> Adding documentation to fits_support seems outside the scope of this PR.
> 

It seems entirely within the scope of this PR to me (in hindsight the documentation really should have been part of #110). Since ASDF-in-Fits support of some fashion is being moved specifically to this package meaning it needs to be documented. This PR is intended to add things to help support the transition of ASDF-in-Fits from `asdf` to `stdatamodels`. Documentation of this support is entirely in-scope of these changes IMOP (if you disagree, it still needs to happen and should be a follow-on PR).

> What is your opinion on adding the migration documentation to ASDF as it has more thorough documentation and these functions are being added to aid in migration for ASDF users?

No. the documentation of something, `fits_support` and/or the "helper functions" for the ASDF-in-Fits transition should live directly alongside the code implementing this. Thus the migration documentation should live here. Insofar as the ASDF documentation, we will need to declare that ASDF-in-Fits is deprecated in the documentation, and then link to the migration guide here using an intersphinx link.

**Comment by braingram** [comment]
> It seems entirely within the scope of this PR to me (in hindsight the documentation really should have been part of #110). Since ASDF-in-Fits support of some fashion is being moved specifically to this package meaning it needs to be documented. This PR is intended to add things to help support the transition of ASDF-in-Fits from `asdf` to `stdatamodels`. Documentation of this support is entirely in-scope of these changes IMOP (if you disagree, it still needs to happen and should be a follow-on PR).

A follow-on PR sounds great.
 
> No. the documentation of something, `fits_support` and/or the "helper functions" for the ASDF-in-Fits transition should live directly alongside the code implementing this. Thus the migration documentation should live here. Insofar as the ASDF documentation, we will need to declare that ASDF-in-Fits is deprecated in the documentation, and then link to the migration guide here using an intersphinx link.

I will work on a module level docstring for asdf_in_fits that describes the migration. Will this be sufficient for the changes you are requesting or did you have more in mind?

**Comment by WilliamJamieson** [comment]
> I will work on a module level docstring for asdf_in_fits that describes the migration. Will this be sufficient for the changes you are requesting or did you have more in mind?

Sure, though it ought to make it into the docs for `stdatamodels`. Its fine if this goes into the follow-on PR.

**Comment by braingram** [comment]
> > I will work on a module level docstring for asdf_in_fits that describes the migration. Will this be sufficient for the changes you are requesting or did you have more in mind?
> 
> Sure, though it ought to make it into the docs for `stdatamodels`. Its fine if this goes into the follow-on PR.

The asdf_in_fits module is already included in the stdatamodels docs through the addition in `api.rst`. Should this be added somewhere else?

**Comment by WilliamJamieson** [comment]
> The asdf_in_fits module is already included in the stdatamodels docs through the addition in `api.rst`. Should this be added somewhere else?

I forgot about that. I think that will work, so long as we can create a sphinx link to it.

**Comment by braingram** [comment]
I added a module docstring to asdf_in_fits. @WilliamJamieson input/edits/suggestions for this are appreciated.

I'm also seeing the built docs on readthedocs with this PR. Thanks @zacharyburnett

**Comment by braingram** [comment]
@WilliamJamieson Are there additional changes you are requesting?

**Comment by braingram** [comment]
@zacharyburnett did this miss the 1.0 release? If not, it would be helpful to have this in that release.

@WilliamJamieson pinging you again for clarification on what changes you are requesting.

**Comment by braingram** [comment]
I rebased this and moved the changes to 1.1.0

File: src/stdatamodels/asdf_in_fits.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+Finally providing the tree and hdulist to
+:func:`stdatamodels.asdf_in_fits.write`. ::
+
+    stdatamodels.asdf_in_fits.write('some_file.fits', tree, hdulist)
+
+When read back with :func:`stdatamodels.asdf_in_fits.open` the data for
+``sci`` and ``dq`` will be read from the HDUList instead of from the
+ASDF data embeded in the HDUList.
+"""
```
Comment (WilliamJamieson) [inline src/stdatamodels/asdf_in_fits.py:79]: I would like to see this turned into a proper docs page not just something in the API docs.

Comment (WilliamJamieson) [reply src/stdatamodels/asdf_in_fits.py:79]: To elaborate, it is not obvious where to find this in the docs and it is just part of the API page like we are trying to obfuscate it. It should be its own page say `asdf_in_fits.rst` and should be linked to on the main page of the docs.

File: docs/source/index.rst [hunk 3]
```diff
@@ -15,6 +15,13 @@ JWST
    jwst/datamodels/index.rst
    jwst/transforms/index.rst
 
+AsdfInFits
+==========
+
+.. toctree::
+   :maxdepth: 1
+
+   asdf_in_fits.rst
```
Comment (WilliamJamieson) [inline docs/source/index.rst:24]: This should be moved up to the first `toctree` on line 4.

File: src/stdatamodels/asdf_in_fits.py (hunk 2 above)
Comment (braingram) [reply src/stdatamodels/asdf_in_fits.py]: I moved the description to the docs.
https://stdatamodels--114.org.readthedocs.build/en/114/

File: docs/source/asdf_in_fits.rst [hunk 4]
```diff
@@ -0,0 +1,77 @@
+:py:mod:`stdatamodels.asdf_in_fits` contains functions to help migrate code that
```
Comment (WilliamJamieson) [inline docs/source/asdf_in_fits.rst:1]: Add a title here, and a sphinx reference so that this page can be linked to externally.

**Review by WilliamJamieson — CHANGES_REQUESTED** [review]

File: docs/source/index.rst (hunk 3 above)
Comment (braingram) [reply docs/source/index.rst]: I moved asdf_in_fits.rst to the other toctree and added a title and link.
https://stdatamodels--114.org.readthedocs.build/en/114/
Let me know if these were not the changes you were requesting.

**Review by WilliamJamieson — APPROVED** [review]

**Comment by braingram** [comment]
The jwst ci job failures that appeared in this PR after updating it:
https://github.com/spacetelescope/stdatamodels/actions/runs/4177420337/jobs/7236004607#step:12:1100
are different than those seen when run against master:
https://github.com/spacetelescope/stdatamodels/actions/runs/4161750811/jobs/7236042033#step:12:622
and will require more investigation.

**Comment by zacharyburnett** [comment]
It looks like this is just a truth value issue:
https://github.com/spacetelescope/stdatamodels/actions/runs/4161750811/jobs/7249375601#step:12:321
```
E           AssertionError: 
E           Not equal to tolerance rtol=1e-05, atol=1e-05
E           
E           Mismatched elements: 1 / 3 (33.3%)
E           Max absolute difference: 16.3400141
E           Max relative difference: 0.44103293
E            x: array([[20.709406,  0.46572 ,  4.686621]], dtype=float32)
E            y: array([[37.04942 ,  0.46572 ,  4.686621]])
```

**Comment by braingram** [comment]
> It looks like this is just a truth value issue: https://github.com/spacetelescope/stdatamodels/actions/runs/4161750811/jobs/7249375601#step:12:321

Thanks for taking a look. I think the run you linked was run against master. The one against this PR has extra errors:
https://github.com/spacetelescope/stdatamodels/actions/runs/4177420337/jobs/7236004607#step:12:1100
```
  FAILED .tox/test-jwst-xdist-cov/lib/python3.11/site-packages/jwst/stpipe/tests/test_entry_points.py::test_get_steps
  FAILED .tox/test-jwst-xdist-cov/lib/python3.11/site-packages/jwst/stpipe/tests/test_step.py::test_step_from_commandline_class_alias
  FAILED .tox/test-jwst-xdist-cov/lib/python3.11/site-packages/jwst/stpipe/tests/test_step.py::test_step_from_commandline_config_class_alias
  FAILED .tox/test-jwst-xdist-cov/lib/python3.11/site-packages/jwst/stpipe/tests/test_utilities.py::test_resolve_step_class_alias
```
Are those also because of the version issues?

**Comment by braingram** [comment]
No class names were changed in this PR.

I think the crds caching isn't configured correctly anymore. The key is `crds-` instead of something like `crds-jwst_1045.pmap`
https://github.com/spacetelescope/stdatamodels/actions/runs/4177420337/jobs/7255362938#step:9:4

I pushed a commit re-adding the `crds-context` id that was removed in 7690332bed152b16e36076afcf5bbb04362bafa8

Any objection if I delete some of the `crds-` caches and possibly other `crds-...pmap` caches to see if this is related?

**Pull request merged**